src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2265,107 @@ private void EndEntry()
             }
         }
 
+        /// <summary>
+        /// For response de-serialization, restore the null values for properties that are omitted
+        /// by the omit-values=nulls preference header.
+        /// </summary>
+        private void RestoreOmittedNullValues()
+        {
+            // Restore omitted null value only when processing response and when
+            // Preference header omit-values=nulls is specified.
+            if (ShouldRestoreNullValues())
+            {
+                IODataJsonLightReaderResourceState resourceState = this.CurrentResourceState;
+                IEdmStructuredType edmStructuredType = resourceState.ResourceType;
+
+                if (resourceState.SelectedProperties == SelectedPropertiesNode.Empty)
+                {
+                    return;
+                }
+                else
+                {
+                    if (resourceState.Resource != null && edmStructuredType != null)
+                    {
+                        ODataResourceBase resource = resourceState.Resource;
+
+                        IEnumerable<IEdmProperty> selectedProperties = resourceState.SelectedProperties.GetSelectedProperties(edmStructuredType);
+
+                        // All dynamic properties: null value restoration for properties that are not present.
+                        foreach (string name in resourceState.SelectedProperties.GetSelectedDynamicProperties(edmStructuredType))
+                        {
+                            if (resource.Properties.Any(p => p.Name.Equals(name, StringComparison.Ordinal))
+                                || resourceState.NavigationPropertiesRead.Contains(name))
+                            {
+                                continue;
+                            }
+
+                            RestoreNullODataProperty(name, resourceState);
+                        }
+
+                        foreach (IEdmProperty currentProperty in selectedProperties)
+                        {
+                            Debug.Assert(currentProperty.Type != null, "currentProperty.Type != null");
+
+                            // Skip declared properties that are not null-able types.
+                            // Skip navigation properties (no navigation links need to be restored as null.)
+                            // Skip properties that have been read (primitive or structural).
+                            if (!currentProperty.Type.IsNullable
+                                || currentProperty.PropertyKind == EdmPropertyKind.Navigation
+                                || resource.Properties.Any(p => p.Name.Equals(currentProperty.Name, StringComparison.Ordinal))
+                                || resourceState.NavigationPropertiesRead.Contains(currentProperty.Name))

```

Comment:
>resourceState.NavigationPropertiesRead.Contains(currentProperty.Name) [](start = 35, length = 69)

if you already excluded navigation properties, why would you have to look in NavigationPropertiesRead?

Reply:
As discussed, we need to check NavigationPropertiesRead for structural property (which is not nav link) in order to skip its null restoration as required.

---
In reply to: [229536601](https://github.com/OData/odata.net/pull/1106#discussion_r229536601) [](ancestors = 229536601)